k8s: Add support for Kubernetes 1.30.0

christarazi opened the pull request:
Update to k8s 1.30.0 and fix code-generator scripts.

Also, bump sigs.k8s.io/controller-runtime to the latest main version as
of this commit (v0.17.1-0.20240327193027-21368602d84b) to workaround a
breaking change introduced in 1.30 libraries.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Add support for Kubernetes 1.30.0

Update to k8s 1.30.0 and fix code-generator scripts.

Also, bump sigs.k8s.io/controller-runtime to the latest main version as
of this commit (v0.17.1-0.20240327193027-21368602d84b) to workaround a
breaking change introduced in 1.30 libraries.

Co-authored-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on .github/actions/aws/k8s-versions.yaml:
```diff
@@ -6,15 +6,17 @@ include:
     region: ca-central-1
     ipsec: true
-  - version: "1.29"
+  - version: "1.30"
```
I don't think this will work because this version is not available yet.

aanm commented on .github/actions/ginkgo/main-focus.yaml:
```diff
@@ -217,91 +217,91 @@ include:
+  - k8s-version: "1.29"
     focus: "f14-datapath-service-ns-xdp-2"
 
   # These tests require an external node which is only available on 1.28
```
```suggestion
  # These tests require an external node which is only available on 1.29
```

aanm commented on .github/actions/ginkgo/main-k8s-versions.yaml:
```diff
@@ -1,28 +1,28 @@
+  - k8s-version: "1.30"
     ip-family: "dual"
     # renovate: datasource=docker
     kube-image: "kindest/node:v1.29.2@sha256:51a1434a5397193442f0be2a297b488b6c919ce8a3931be0ce822606ea5ca245"
```
```suggestion
    kube-image: "quay.io/cilium/kindest-node:v1.30.0-rc.0@sha256:98fff7e882a1c93e9d7e06143476e982a2916009182c7f019eb95e5cda7cda0d"
```

aanm commented on .github/actions/ginkgo/main-k8s-versions.yaml:
```diff
@@ -1,28 +1,28 @@
+  - k8s-version: "1.29"
     ip-family: "dual"
     # renovate: datasource=docker
     kube-image: "kindest/node:v1.28.7@sha256:9bc6c451a289cf96ad0bbaf33d416901de6fd632415b076ab05f5fa7e4f65c58"
```
```suggestion
    kube-image: "kindest/node:v1.29.2@sha256:51a1434a5397193442f0be2a297b488b6c919ce8a3931be0ce822606ea5ca245"
```

aanm commented on .github/actions/ginkgo/main-k8s-versions.yaml:
```diff
@@ -1,28 +1,28 @@
+  - k8s-version: "1.28"
     ip-family: "dual"
     # renovate: datasource=docker
     kube-image: "kindest/node:v1.27.11@sha256:681253009e68069b8e01aad36a1e0fa8cf18bb0ab3e5c4069b2e65cafdd70843"
```
```suggestion
    kube-image: "kindest/node:v1.28.7@sha256:9bc6c451a289cf96ad0bbaf33d416901de6fd632415b076ab05f5fa7e4f65c58"
```

aanm commented on .github/actions/ginkgo/main-k8s-versions.yaml:
```diff
@@ -1,28 +1,28 @@
+  - k8s-version: "1.27"
     ip-family: "dual"
     # renovate: datasource=docker
     kube-image: "kindest/node:v1.26.14@sha256:5d548739ddef37b9318c70cb977f57bf3e5015e4552be4e27e57280a8cbb8e4f"
```
```suggestion
  - k8s-version: "1.27"
    ip-family: "dual"
    # renovate: datasource=docker
    kube-image: "kindest/node:v1.27.11@sha256:681253009e68069b8e01aad36a1e0fa8cf18bb0ab3e5c4069b2e65cafdd70843"
    # renovate: datasource=docker depName=quay.io/lvh-images/kind
    kernel: "5.4-20240305.092417@sha256:054026744db7305801641f276d751d6179dc5eed9a1eaa4c8b6918982e5f828e"
```

aanm commented on Makefile:
```diff
@@ -323,29 +323,26 @@ generate-hubble-api: api/v1/flow/flow.proto api/v1/peer/peer.proto api/v1/observ
 
 define generate_k8s_api
```
Do we need these changes? I know that kube_codegen.sh was introduced but if the older script it's still available let's use it. Changeing to the `kube_codegen.sh` will require more changes IIRC.

sayboras replied:
The old script is available, but didn't do anything right ? :thinking: 

https://github.com/kubernetes/code-generator/blob/v0.30.0-rc.1/generate-internal-groups.sh

> echo "ERROR: $(basename "$0") has been removed."
echo "ERROR: Please use k8s.io/code-generator/kube_codegen.sh instead."

christarazi replied:
Yes it was removed and we had to change the scripts to use `kube_codegen.sh`.

aanm commented on test/provision/k8s_install.sh:
```diff
@@ -589,6 +589,20 @@ case $K8S_VERSION in
+        sudo apt-get install -y conntrack
+        KUBERNETES_CNI_VERSION="1.1.1"
+        KUBERNETES_CNI_OS="-linux"
+        K8S_FULL_VERSION="1.30.0"
```
This should be `1.30.0-rc.0`

christarazi committed:
k8s/slim: Update readme with gotchas regarding updating slim structures

While updating the slim structures, there are a few gotchas to keep in
mind. If a developer fails to follow the advice, the generated changes
from `make generate-k8s-api` leads to very hard to debug/understand
errors.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/k8s/slim/k8s/apis/meta/v1/types.go:
```diff
@@ -134,7 +146,6 @@ type ObjectMeta struct {
 	// Value must be treated as opaque by clients and .
 	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency
 	// +optional
-	// +deepequal-gen=false
```
Why did you remove the deepequals? :thinking:

christarazi replied:
This was unintended. Re-adding.

aanm commented on pkg/k8s/slim/k8s/apis/meta/v1/types.go:
```diff
@@ -235,11 +267,11 @@ type LabelSelector struct {
 	// operator is "In", and the values array contains only "value". The requirements are ANDed.
-	//
-	// +kubebuilder:validation:Optional
-	MatchLabels map[string]MatchLabelsValue `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
```
I think we need to leave the `MatchLabelsValue` because of the schema validation, is that not necessary anymore?

christarazi replied:
Ah, you're right. I also added that as an extra gotcha in the documention commit.

aanm commented on pkg/k8s/slim/k8s/apis/meta/v1/types.go:
```diff
@@ -278,6 +308,76 @@ const (
 
+// ManagedFieldsEntry is a workflow-id, a FieldSet and the group version of the resource
+// that the fieldset applies to.
+type ManagedFieldsEntry struct {
```
I think we can remove this entire structure since it's not used anywhere.

aanm reviewed (changes requested):
There are somethings I noticed on a first pass. I'll do a second pass tomorrow.

aanm commented on .github/actions/ginkgo/main-focus.yaml:
```diff
@@ -187,7 +187,7 @@ include:
     cliFocus: 'K8sDatapathServicesTest Checks E/W loadbalancing \\(ClusterIP, NodePort from inside cluster, etc\\) TFTP|K8sDatapathServicesTest Checks E/W loadbalancing \\(ClusterIP, NodePort from inside cluster, etc\\) with'
 
   ###
-  # K8sDatapathServicesTest Checks E/W loadbalancing (ClusterIP, NodePort from inside cluster, etc) Tests NodePort inside cluster (kube-proxy) 
```
oh, this space is needed, that's how the test is named. Otherwise the linter will fail.

christarazi replied:
Oh good catch. My editor prunes extra whitespace on save so I didn't intend to do this.

aanm commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwideenvoyconfigs.yaml:
```diff
@@ -93,11 +93,6 @@ spec:
-                          - In
-                          - NotIn
-                          - Exists
-                          - DoesNotExist
```
What happened here?

christarazi replied:
Fixed

aanm commented on pkg/k8s/slim/k8s/apis/meta/v1/types.go:
```diff
@@ -256,15 +289,13 @@ type LabelSelectorRequirement struct {
 	// operator represents a key's relationship to a set of values.
 	// Valid operators are In, NotIn, Exists and DoesNotExist.
-	//
-	// +kubebuilder:validation:Enum=In;NotIn;Exists;DoesNotExist
```
I think this was accidentally removed

christarazi replied:
Yep, my mistake. Another example of something to keep in mind when updating the slim structures. There are lots of these little changes that we've made over the years that we depend on. Now I'm documenting them in the README of the slim directory.

christarazi replied:
See https://github.com/cilium/cilium/compare/af14f15fb803cc01c117206dc43ea057f6ad2ca9..4afcaa19c2852acdc40c125ed2817a15cd08ef49

joestringer commented on Makefile:
```diff
@@ -364,7 +347,7 @@ define generate_k8s_protobuf
 		--packages=$(subst $(newline),$(comma),$(1)) \
 		--go-header-file "$(PWD)/hack/custom-boilerplate.go.txt" \
-		--output-base=$(2)
+		--output-dir="$(PROTOBUF_OUTPUT_DIR)"
```
This macro now receives the `TMPDIR` as a parameter, but ignores it.

Why do we have the TMPDIR at all? I figure it's so that we can use the existing files in the tree to generate a bunch of new files in an entirely separate path, then iff all generation succeeds, we copy the full path back into the main directory. Seems a bit odd to be now generating files somewhere within the current directory instead?

joestringer replied:
Is the base problem here that `go-to-protobuf` removed the parameters we're relying on and refuses to recreate the directory structure in the output dir based on the input directory structure, thereby forcing us to directly inject the files back into the same tree?

joestringer replied:
Ugh yeah looks like upstream broke the ability to output somewhere outside `$GOPATH`. I see other commands adding a `--output-package` to provide the equivalent `$GOPATH` path presumably to infer something about the output, but they didn't bother for this command.

The best I'm coming up with so far as an alternative here is accept dirtying the tree and use this instead:
```
    --output-dir=$$GOPATH/src
```

I'll carry this change in https://github.com/cilium/cilium/pull/32063 since it'll be a dependency there if you don't integrate it in before merge.

joestringer commented on Makefile:
```diff
@@ -390,14 +371,8 @@ generate-k8s-api: ## Generate Cilium k8s API client, deepcopy and deepequal Go s
-	$(QUIET) $(call generate_k8s_api,client$(comma)lister$(comma)informer,github.com/cilium/cilium/pkg/k8s/client,github.com/cilium/cilium/pkg/k8s/apis,$(GEN_CRD_GROUPS),"$(TMPDIR)")
+	$(QUIET) contrib/scripts/k8s-code-gen.sh "$(TMPDIR)"
 
-	$(QUIET) cp -r "$(TMPDIR)/github.com/cilium/cilium/." ./
```
nit: Given that there are various steps inside this Make target which generate files within the TMPDIR, I don't think it makes sense to hide the copy inside the `k8s-code-gen.sh` script.

christarazi replied:
This part was added by @aanm. Andre, is there a reason for moving the `cp` into the script?

aanm replied:
@joestringer @christarazi yes, the `cp` needs to be done before the code generation for the remaining of the packages. You can test it out by removing the `cp` from the script and re-add it in the Makefile to see the difference of the results

joestringer replied:
I see. Probably the solution to my nit is to move even more of the build process out of make, but that'd be a bigger ask. Feels messy to me, but nothing to hold up this PR over.

joestringer reviewed (commented):
Some minor nits and observations from the Make target pieces. I feel like the make targets we have here are fighting with the upstream k8s codegen tooling a bit. Non-blocking feedback, we can iterate further in-tree.

tklauser commented on pkg/k8s/slim/README.md:
```diff
@@ -89,3 +90,24 @@ not optimized yet, you will hit the error:
 make sure the package is registered and the client is supposed to be used
 for that structure.
+
+## Gotchas for updating slim structures
+
+* Remove any changes regarding comments starting with "import ..." because that
+  affects the protoc's code generation
+* Remove any changes regarding a change of the import path for the `core` and
+  `meta` v1* packages, i.e. undo changes that cause the import path to fallback
+  to "k8s.io/..." rather than "pkg/k8s/slim/..."
+* Remove all new structures that are not used by Cilium and only update the
+  structures that are in use
+* Ignore all changes to `pkg/k8s/slim/k8s/apis/meta/v1/meta.go` regarding the
+  `Object` interface methods, unless the updating you are performing requires
+  these interface methods to change.
+* Ignore changes that remove that the `kubebuilder` directives on the
+  `LabelSelectorRequirement` struct, especially for the `Operator` field.
+* Ignore changes to `LabelSelector` inside
+  `pkg/k8s/slim/k8s/apis/meta/v1/types.go` and `MatchLabelsValue`. The latter
+  is a custom type that we depend on.
+* Ignore changes inside `pkg/k8s/slim/k8s/apis/labels/selector.go` regarding
+  new fields in the `Parser` struct, especially with `path`. If new fields are
+  added to the struct, add them if you know what you are doing.
```
Thanks for adding this section :heart:

ysksuzuki reviewed (approved):
CEGP CRD change looks good to me
